src/parse.rs

```diff
@@ -184,70 +184,57 @@ impl<'a> Bytes<'a> {
             return Err(Error::ExpectedInteger);
         }
 
-        let s = unsafe { from_utf8_unchecked(&self.bytes[0..num_bytes]) };
-
-        if s.as_bytes()[0] == b'_' {
+        if self.bytes[0] == b'_' {
             return Err(Error::UnderscoreAtBeginning);
         }
 
         fn calc_num<T: Num>(
-            bytes: &Bytes,
-            s: &str,
+            bytes: &mut Bytes,
+            b: &[u8],
             base: u8,
             mut f: impl FnMut(&mut T, u8) -> bool,
         ) -> Result<T> {
             let mut num_acc = T::from_u8(0);
 
-            for &byte in s.as_bytes() {
+            for (i, &byte) in b.iter().enumerate() {
                 if byte == b'_' {
                     continue;
                 }
 
                 if num_acc.checked_mul_ext(base) {
+                    let _ = bytes.advance(b.len());
                     return Err(Error::IntegerOutOfBounds);
                 }
 
                 let digit = bytes.decode_hex(byte)?;
 
                 if digit >= base {
-                    return Err(Error::ExpectedInteger);
+                    let _ = bytes.advance(i);
+                    return Err(Error::InvalidIntegerDigit {
+                        digit: char::from(byte),
+                        base,
+                    });
                 }
 
                 if f(&mut num_acc, digit) {
+                    let _ = bytes.advance(b.len());
                     return Err(Error::IntegerOutOfBounds);
                 }
             }
 
+            let _ = bytes.advance(b.len());
+
             Ok(num_acc)
         }
 
-        let res = if sign > 0 {
-            calc_num(self, s, base, T::checked_add_ext)
+        if sign > 0 {
+            calc_num(self, &self.bytes[0..num_bytes], base, T::checked_add_ext)
         } else {
-            calc_num(self, s, base, T::checked_sub_ext)
-        };
-
-        let _ = self.advance(num_bytes);
-
-        res
-    }
-
-    pub fn any_number(&mut self) -> Result<Number> {
-        fn any_float(f: f64) -> Result<Number> {
-            // total_cmp ensures that NANs are treated properly
-            if f.total_cmp(&f64::from(f as f32)).is_eq() {
-                Ok(Number::F32((f as f32).into()))
-            } else {
-                Ok(Number::F64(f.into()))
-            }
-        }
-
-        let bytes_backup = self.bytes;
-
-        if self.next_bytes_is_float() {
-            return any_float(self.float::<f64>()?);
+            calc_num(self, &self.bytes[0..num_bytes], base, T::checked_sub_ext)
         }
+    }
 
+    pub fn any_integer(&mut self, desire: DesireInteger) -> Result<ParsedInteger> {
```

Comment:
Could this be refactored to again take a generic with a trait that takes care of implementing the parsing and unwrapping from a parsed int, same for float, this should allow moving the error handling to parse and using the literal in the error message